l1/l1.go

```diff
@@ -0,0 +1,141 @@
+package l1
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/l1/internal/contract"
+	"github.com/NethermindEth/juno/service"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+//go:generate mockgen -destination=./mocks/mock_subscriber.go -package=mocks github.com/NethermindEth/juno/l1 Subscriber
+type Subscriber interface {
+	WatchHeader(ctx context.Context, sink chan<- *types.Header) (event.Subscription, error)
+	WatchLogStateUpdate(ctx context.Context, sink chan<- *contract.StarknetLogStateUpdate) (event.Subscription, error)
+}
+
+type Client struct {
+	l1                Subscriber
+	l2Chain           *blockchain.Blockchain
+	log               utils.SimpleLogger
+	confirmationQueue *queue
+}
+
+var _ service.Service = (*Client)(nil)
+
+func NewClient(l1 Subscriber, chain *blockchain.Blockchain, confirmationPeriod uint64, log utils.SimpleLogger) *Client {
+	return &Client{
+		l1:                l1,
+		l2Chain:           chain,
+		log:               log,
+		confirmationQueue: newQueue(confirmationPeriod),
+	}
+}
+
+func (c *Client) subscribeToHeaders(ctx context.Context, buffer int) (event.Subscription, chan *types.Header, error) {
+	headerChan := make(chan *types.Header, buffer)
+	sub, err := c.l1.WatchHeader(ctx, headerChan)
+	if err != nil {
+		return nil, nil, fmt.Errorf("subscribe to L1 headers: %w", err)
+	}
+	return sub, headerChan, nil
+}
+
+func (c *Client) subscribeToUpdates(ctx context.Context, buffer int) (event.Subscription, chan *contract.StarknetLogStateUpdate, error) {
+	logStateUpdateChan := make(chan *contract.StarknetLogStateUpdate, buffer)
+	sub, err := c.l1.WatchLogStateUpdate(ctx, logStateUpdateChan)
+	if err != nil {
+		return nil, nil, fmt.Errorf("subscribe to L1 state updates: %w", err)
+	}
+	return sub, logStateUpdateChan, nil
+}
+
+func (c *Client) Run(ctx context.Context) error {
+	buffer := 128
+	subscribeCtx, subscribeCtxCancel := context.WithCancel(ctx)
+	defer subscribeCtxCancel()
+
+	subUpdates, logStateUpdateChan, err := c.subscribeToUpdates(subscribeCtx, buffer)
+	if err != nil {
+		return err
+	}
+	defer subUpdates.Unsubscribe()
+
+	subHeaders, headerChan, err := c.subscribeToHeaders(subscribeCtx, buffer)
+	if err != nil {
+		return err
+	}
+	defer subHeaders.Unsubscribe()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-subHeaders.Err():
+			c.log.Warnw("L1 header subscription failed, resubscribing", "error", err)
+			subHeaders.Unsubscribe()
+
+			subHeaders, headerChan, err = c.subscribeToHeaders(subscribeCtx, buffer)
+			if err != nil {
+				return err
+			}
+			defer subHeaders.Unsubscribe() //nolint:gocritic
```

Comment:
> will refer to the same subHeader

We reassign it a few lines above? If I'm missing something, please let me know.

Reply:
Here is an example:

```go
import "fmt"

func main() {
	x := 2
	defer add2(&x)

	x = 0

	// This program will print the following:
	// 2 (not 4, as x was reassigned before it could be evaluated,
        // It will print 4 if add2 accepted a value instead of a reference)
}

func add2(x *int) {
	*x += 2
	fmt.Println(*x)
}
```

Therefore, when the line 68 `Unsubscribe()` method is called it will refer to the subHeader which was most recently assigned, not the initial one. And since you explicitly `Unsubscribe()` if an error occurs. I thought that line 88 was redundant.

Reply:
After reading the documentation, `Unsubscribe()` can be called any number of times, so there won't be a panic.